sql/core/src/test/scala/org/apache/spark/sql/SubquerySuite.scala

```diff
@@ -204,6 +204,30 @@ class SubquerySuite extends QueryTest with SharedSparkSession {
     }
   }
 
+  test("SPARK-29145: JOIN Condition use QueryList") {
+    withTempView("s1", "s2", "s3") {
+      Seq(1, 3, 5, 7, 9).toDF("id").createOrReplaceTempView("s1")
+      Seq(1, 3, 4, 6, 9).toDF("id").createOrReplaceTempView("s2")
+      Seq(3, 4, 6, 9).toDF("id").createOrReplaceTempView("s3")
+
+      checkAnswer(
+        sql("SELECT s1.id from s1 JOIN s2 ON s1.id = s2.id and s1.id IN (select 9)"),
+        Row(9) :: Nil)
+
+      checkAnswer(
+        sql("SELECT s1.id from s1 JOIN s2 ON s1.id = s2.id and s1.id NOT IN (select 9)"),
+        Row(1) :: Row(3) :: Nil)
+
+      checkAnswer(
+        sql("SELECT s1.id from s1 JOIN s2 ON s1.id = s2.id and s1.id IN (select id from s3)"),
```

Comment:
for example, do we support
`SELECT s1.id from s1 JOIN s2 ON s1.id = s2.id and s1.id IN (select id from s3 where s3.id = s2.id)`

Reply:
> for example, do we support
> `SELECT s1.id from s1 JOIN s2 ON s1.id = s2.id and s1.id IN (select id from s3 where s3.id = s2.id)`

Cann't since strategy's idempotence is broken.  Seem write sql like this is not reasonable...

Reply:
also cc @dilipbiswal 

I checked with pgsql and it's supported. We need to update `RewriteCorrelatedScalarSubquery` to support it in Spark.

Reply:
> also cc @dilipbiswal
> 
> I checked with pgsql and it's supported. We need to update `RewriteCorrelatedScalarSubquery` to support it in Spark.

We should support it, checking on this issue.

Reply:
We need to address the support in this pr? I think its ok to do in another jira. kindly ping @dilipbiswal